src/main/java/org/wildfly/security/keystore/KeyStoreWatcher.java

```diff
@@ -0,0 +1,209 @@
+                while (storeIterator.hasNext()) {
+                    Store current = storeIterator.next();
+                    if (keyStore == current) {
+                        storeIterator.remove();
+                    }
+                }
+                if (pathStores.isEmpty()) {
+                    pathRegistration.remove(fileName);
+                } else {
+                    pathRegistration.put(fileName, pathStores);
+                }
+            }
+
+            Map<Path, Map<String, List<Store>>> newWatchedPaths = new HashMap<Path, Map<String, List<Store>>>(watchedPaths);
+            if (pathRegistration.isEmpty()) {
+                newWatchedPaths.remove(dirPath);
+                WatchKey key = registrations.remove(dirPath);
+                if (key != null) {
+                    key.cancel();
+                }
+                if (newWatchedPaths.isEmpty()) {
+                    watchService.close();
+                    watchService = null;
+                }
+            } else {
+                newWatchedPaths.put(dirPath, pathRegistration);
+            }
+            watchedPaths = newWatchedPaths;
+        }
+
+    }
+
+    interface Store {
+
+        void modified();
+
+    }
+
+    private class EventTaker implements Runnable {
+
+        @Override
+        public void run() {
+            try {
+                WatchService watchService = KeyStoreWatcher.this.watchService;
+                while (watchService != null) {
+                    WatchKey key = watchService.take();
+                    Path watchedPath = (Path) key.watchable();
+                    Map<String, List<Store>> pathRegistration = watchedPaths.get(watchedPath);
+                    Set<Store> toNotify = new HashSet<Store>();
+                    if (pathRegistration != null) {
+                        for (WatchEvent<?> event : key.pollEvents()) {
+                            if (StandardWatchEventKinds.ENTRY_CREATE.equals(event.kind())
+                                    || StandardWatchEventKinds.ENTRY_MODIFY.equals(event.kind())) {
+                                Path context = (Path) event.context();
+                                String name = context.getFileName().toString();
+                                System.out.println("File Name " + name);
+                                List<Store> stores = pathRegistration.get(name);
+                                if (stores != null) {
+                                    for (Store current : stores) {
+                                        toNotify.add(current);
+                                    }
+                                }
+
+                            } else if (StandardWatchEventKinds.OVERFLOW.equals(event.kind())) {
+                                // No idea what happened so reload them all.
+                                for (List<Store> stores : pathRegistration.values()) {
+                                    for (Store current : stores) {
```

Comment:
I have clarified it in the next commit, the two methods that modify the collections are modifying copies - just added wrapping to guarantee they are not modifiable.